### ISSUE TITLE
Windows Build broken

### ISSUE THREAD

**Issue by mciprian13** [pytorch/glow#4712]
The Windows build is broken with the following message:
[glow_build.log](https://github.com/pytorch/glow/files/4949497/glow_build.log)
Most likely due to #4699.

**Comment by mciprian13** [comment]
@opti-mix @ayermolo Do you know a solution to this? Thanks!

**Comment by opti-mix** [comment]
OK. It looks like Microsoft Visual Studio does support packed and aligned attributes. Instead, they have a different syntax described here:
https://docs.microsoft.com/en-us/cpp/preprocessor/pack?view=vs-2019

Probably `#pragma pack(2)` is needed for bfloat16 class definition, when it is compiled on Windows. We don't have any possibilities to test on Windows here, but it should be pretty straight forward for one of you to test if this approach works and submit a PR to fix it. Thanks!

**Comment by ayermolo** [comment]
@mciprian13 sorry, I no longer have access to windows machine.

**Comment by mciprian13** [comment]
@opti-mix I tried doing this:
```
#ifdef WIN32
#pragma pack(2)
class bfloat16 {
#else
class __attribute__((packed, aligned(2))) bfloat16 {
#endif
```
but now the Glow build fails due to some LLVM header files (currently using 8.0):
```
"F:\glow-vcc-build\glow_build\ALL_BUILD.vcxproj" (default target) (1) ->
"F:\glow-vcc-build\glow_build\lib\Backends\Backends.vcxproj" (default target) (23) ->
"F:\glow-vcc-build\glow_build\lib\Backends\CPU\CPUBackend.vcxproj" (default target) (24) ->
"F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj" (default target) (30) ->
(ClCompile target) -> 
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/ADT/PointerSumType.h(256): error C2338: This discriminant value requires too many bits! [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/ADT/PointerSumType.h(252): error C2338: This discriminant value requires too many bits! [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
  F:\glow-vcc-build\llvm\llvm_80\llvm\include\llvm/IR/DerivedTypes.h(146): error C2338: Alignment sufficient for objects appended to FunctionType [F:\glow-vcc-build\glow_build\lib\LLVMIRCodeGen\LLVMIRCodeGen.vcxproj]
```
Is there some other problem in LLVMIRCodeGen?

**Comment by opti-mix** [comment]
@mciprian13 Yeah, the joy of using Visual Studio ;-) Thanks for trying it out!

Please have a look at the `alignas` description:
https://docs.microsoft.com/en-us/cpp/cpp/alignment-cpp-declarations?view=vs-2019
If this one works properly, and creates a struct of size 2 and aligned at 2 bytes that would be ideal. Then we don't need anything else.

And how to use `#pragma pack` correctly:
https://stackoverflow.com/questions/31034906/scope-of-the-pragma-pack-directive-in-visual-studio
It seems like we need something like:
```c++
#pragma pack(push, 2)
struct/class definition
#pragma pack(pop)
```
Otherwise the effect of the first `#pragma pack` lasts forever! ;-)

**Comment by mciprian13** [comment]
@opti-mix I tried in #4719 using the `alignas` and the build works for both Linux/Windows worlds.
I guess the static assert from `BFloat16.h` verifies that the size is truly 2 bytes:
```
using bfloat16_t = bfloat16;
static_assert(sizeof(bfloat16_t) == 2, "bfloat16_t must be 16 bits wide");
```
right? Is there something else to verify?

**Comment by mciprian13** [comment]
@opti-mix Thanks a lot for the suggestion. The Glow build now works for everybody 😄
Can you also look over #4602 which is something we need on our side but was also reported here on the community. Thanks!

**Comment by mciprian13** [comment]
Solved in #4719. Closing this issue.
